### PR TITLE
Add a regression test for partial unfuse_lora bookkeeping

### LINKED ISSUE #14214: unfuse_lora(components=[...]) desyncs _merged_adapters/num_fused_loras from actual per-component PEFT merge state
### Describe the bug

`LoraBaseMixin._merged_adapters` (the set backing `num_fused_loras`/`fused_loras`) is a single set shared across the whole pipeline, but `fuse_lora()`/`unfuse_lora()` both accept a `components` argument that operates on a subset of `_lora_loadable_modules` (e.g. `unet`, `text_encoder`) independently. When the same adapter is fused into multiple components (the default, since `fuse_lora()` with no `adapter_names` fuses into every requested component) and later only *some* of those components are unfused via `unfuse_lora(components=[...])`, `_merged_adapters` drops the adapter name entirely, even though it's still physically merged into the base weights of the untouched component(s).

`num_fused_loras`/`fused_loras` then report the pipeline as having nothing (or less) fused, while the untouched component's weights still silently contain the baked-in LoRA delta. If I intend to submit a PR for this, and have one ready.

### Reproduction

```python
import torch.nn as nn
from peft import LoraConfig
from diffusers.loaders.lora_base import LoraBaseMixin
from diffusers.loaders.peft import PeftAdapterMixin
from diffusers.models.modeling_utils import ModelMixin
from diffusers.configuration_utils import ConfigMixin


class TinyModel(ModelMixin, ConfigMixin, PeftAdapterMixin):
    config_name = "config.json"

    def __init__(self):
        super().__init__()
        self.linear = nn.Linear(8, 8)


class FakePipeline(LoraBaseMixin):
    _lora_loadable_modules = ["unet", "text_encoder"]

    def __init__(self, unet, text_encoder):
        self._merged_adapters = set()
        self.unet, self.text_encoder = unet, text_encoder


unet, text_encoder = TinyModel(), TinyModel()
lora_config = LoraConfig(r=4, lora_alpha=4, target_modules=["linear"], init_lora_weights=False)
unet.add_adapter(lora_config, adapter_name="my_adapter")
text_encoder.add_adapter(lora_config, adapter_name="my_adapter")
pipe = FakePipeline(unet, text_encoder)

pipe.fuse_lora(components=["unet", "text_encoder"], adapter_names=["my_adapter"])
print("after fuse (both):", pipe.num_fused_loras, pipe.fused_loras)

pipe.unfuse_lora(components=["text_encoder"])  # only unfuse text_encoder
print("after unfuse (text_encoder only):", pipe.num_fused_loras, pipe.fused_loras)

from peft.tuners.tuners_utils import BaseTunerLayer
unet_still_merged = any(isinstance(m, BaseTunerLayer) and len(m.merged_adapters) > 0 for m in unet.modules())
print("unet is ACTUALLY still merged at the PEFT level:", unet_still_merged)
```

Output:
```
after fuse (both): 1 {'my_adapter'}
after unfuse (text_encoder only): 0 set()
unet is ACTUALLY still merged at the PEFT level: True
```

`num_fused_loras`/`fused_loras` report `0`/`set()` even though `unet` still has `my_adapter` merged into its base weights.

This is a real, documented, publicly-supported code path, every concrete pipeline mixin (e.g. `StableDiffusionLoraLoaderMixin.unfuse_lora`) exposes the `components` parameter, and the docstring for `unfuse_lora` documents it as "list of LoRA-injectable components to unfuse LoRA from," implying a subset is a normal, intended use.

I also checked the existing test suite (`tests/lora/utils.py`) and none of the `fuse_lora`/`unfuse_lora` tests exercise a partial-`components` combination, they always pass every loadable component at once (`components=self.pipeline_class._lora_loadable_modules`), which is why this gap wasn't caught.

### Logs

_No response, this is a silent correctness/bookkeeping bug, not a crash._

### System Info

Reproduced against the current `main` branch source (`src/diffusers/loaders/lora_base.py`), `peft` 0.19.1, `torch` 2.x, Python 3.12.10, Windows 11. No GPU needed to reproduce, the bug is in pure Python set bookkeeping.

### Who can help?

@sayakpaul (LoRA/PEFT integration maintainer, per the issue template's guidance for the "LoRA" area)

### PR DESCRIPTION
**Pull request by ErenAta16**
## What this is now

The fix this PR originally carried is no longer needed: an equivalent one has landed on `main`. `unfuse_lora` now collects what is still merged across every loadable component and intersects, rather than dropping an adapter on the first component it unmerges from (`src/diffusers/loaders/lora_base.py`):

```python
# Only remove an adapter from _merged_adapters once it is no longer
# physically merged in any remaining loadable component.
remaining_merged: set[str] = set()
for component_name in self._lora_loadable_modules:
    component_model = getattr(self, component_name, None)
    if isinstance(component_model, nn.Module):
        for module in component_model.modules():
            if isinstance(module, BaseTunerLayer):
                remaining_merged.update(module.merged_adapters)
self._merged_adapters = self._merged_adapters & remaining_merged
```

So I have dropped my implementation and reduced this to the regression test. It is now a single file, `+40/-0`.

## Why the test is still worth having

Nothing in the suite exercises a partial unfuse. Every `unfuse_lora` call in `tests/lora/utils.py` passes the full set:

```
utils.py:954    pipe.unfuse_lora(components=self.pipeline_class._lora_loadable_modules)
utils.py:1688   pipe.unfuse_lora(components=self.pipeline_class._lora_loadable_modules)
utils.py:1711   pipe.unfuse_lora(components=self.pipeline_class._lora_loadable_modules)
```

With all components unfused at once, the old code and the new code agree, so the bug was invisible to the suite and would be invisible to it again. `test_fuse_unfuse_partial_components_keeps_merged_adapter_bookkeeping` fuses an adapter into the text encoder and the denoiser, unfuses only the text encoder, and asserts that `num_fused_loras` is still 1 and that the adapter is still in `fused_loras`, then unfuses the denoiser and asserts it drops to 0.

## How far I verified it

I could not run the LoRA suite on this machine, so this is not a full green run and I am not claiming one. What I did do is transcribe the bookkeeping block above verbatim and drive it with stub components standing in for `BaseTunerLayer`, to check the assertions match what the current implementation produces:

```
after fuse on both           num_fused_loras = 1
after unfuse(text_encoder)   num_fused_loras = 1   fused = ['adapter-1']
after unfuse(transformer)    num_fused_loras = 0   fused = []
```

That is the sequence the test asserts. I also checked that everything it depends on is still present in `tests/lora/utils.py` on `main` (`supports_text_encoder_loras`, `check_if_lora_correctly_set`, `get_dummy_components`, `unet_kwargs`, `num_fused_loras`, `fused_loras`), and it is inserted at the same place as before, between `test_simple_inference_with_text_lora_denoiser_fused_multi` and `test_lora_scale_kwargs_match_fusion`.

The branch has been merged up to `main` (`360bef80`) and is mergeable again.

If you would rather not carry a test for an already-fixed bug, closing this is a perfectly good outcome and I will not take it as a loss.

## A practical note

My account cannot post comments on this repository, so updates from me arrive as description edits and commits rather than replies.

### CODE REVIEW

File: tests/lora/utils.py [hunk 1]
```diff
@@ -1711,6 +1711,45 @@ def test_simple_inference_with_text_lora_denoiser_fused_multi(
...
 
+    def test_fuse_unfuse_partial_components_keeps_merged_adapter_bookkeeping(self):
+        """
+        `_merged_adapters` (backing `num_fused_loras`/`fused_loras`) is shared across the whole pipeline,
+        while actual merge state is tracked per component by PEFT. Fusing an adapter into every loadable
+        component and then unfusing only *some* of them should not report the adapter as fully unfused,
+        it's still merged into the untouched component(s).
+        """
+        if "text_encoder" not in self.pipeline_class._lora_loadable_modules:
+            return
```
Comment (sayakpaul) [inline tests/lora/utils.py]: We should use `skip` rather than returning so that the test properly gets skipped.

**Comment by ErenAta16** [comment]
Process note, read after opening this: I found `.ai/AGENTS.md`'s coordination guidance (added in #14185, a couple days before this PR) after the fact, not before, so this didn't go through the "wait for explicit maintainer acknowledgment on the issue before opening a PR" step it asks for. Issue #14214 has the full trace and reproduction but no prior acknowledgment. Apologies for the process miss, sharing the self-review now to at least close that part of the gap.

**Self-review against `.ai/review-rules.md`:**

- **Ephemeral context**: no PR-review-specific comments, no debug printouts, no scratch/dev-path scripts in the diff. The two added comments (on the `unmerge_candidates` tracking and the `_is_adapter_merged_in_any_component` docstring) explain the underlying design tension, self-contained, not "per discussion" references.
- **Documentation impact**: `num_fused_loras`/`fused_loras`'s existing docstrings ("Returns the number of LoRAs that have been fused" / "Returns names of the LoRAs that have been fused") already describe the correct, intended contract, this PR fixes the implementation to actually match what was already documented, not a new or changed public surface. No doc updates needed.
- **Dead code analysis**: not applicable, this isn't a new-model PR.

One thing I did *not* fix and want to flag rather than silently leave: `unfuse_lora`'s existing docstring doesn't mention that `_merged_adapters`/`num_fused_loras` is pipeline-wide rather than per-component, so a caller reading only the docstring wouldn't know that fusing into multiple components and unfusing just one still shows the adapter as fused. Happy to add a short note there if that'd be useful, left it out of this PR to keep the diff scoped to the actual bug rather than bundling in a docstring rewrite, but can add it as a follow-up commit here if you'd rather have it in the same PR.

File: tests/lora/utils.py (hunk 1 above)
Comment (ErenAta16) [reply tests/lora/utils.py]: Good catch, fixed in cacdeee. The old `return` was reporting a pass for pipelines that don't support text encoder LoRAs at all, instead of skipping. Switched to `self.supports_text_encoder_loras` + `pytest.skip(...)`, matching the convention already used elsewhere in this file.

File: tests/lora/utils.py [hunk 2]
```diff
@@ -1711,6 +1711,45 @@ def test_simple_inference_with_text_lora_denoiser_fused_multi(
...
         self.assertTrue(pipe.num_fused_loras == 0, f"{pipe.num_fused_loras=}, {pipe.fused_loras=}")
 
+    def test_fuse_unfuse_partial_components_keeps_merged_adapter_bookkeeping(self):
+        """
+        `_merged_adapters` (backing `num_fused_loras`/`fused_loras`) is shared across the whole pipeline,
+        while actual merge state is tracked per component by PEFT. Fusing an adapter into every loadable
+        component and then unfusing only *some* of them should not report the adapter as fully unfused,
+        it's still merged into the untouched component(s).
+        """
+        if "text_encoder" not in self.pipeline_class._lora_loadable_modules:
```
[the remark this answers is not in the document]
Comment (ErenAta16) [reply tests/lora/utils.py]: Thanks, I looked at this again against AGENTS.md's single-caller inlining guidance. I went back and forth, but decided to leave `_is_adapter_merged_in_any_component` as a separate method: inlining it means a two-level nested loop with a boolean flag sitting inside the `unfuse_lora` unmerge-candidate loop, which reads worse than a named, docstring'd predicate, even with one caller. Happy to inline it if you'd still prefer that on a second look, but wanted to explain the call rather than silently leave it.

File: src/diffusers/loaders/lora_base.py [hunk 3]
```diff
@@ -673,11 +680,28 @@ def unfuse_lora(self, components: list[str] | None = None, **kwargs):
...
+            if not self._is_adapter_merged_in_any_component(adapter):
+                self._merged_adapters = self._merged_adapters - {adapter}
+
+    def _is_adapter_merged_in_any_component(self, adapter_name: str) -> bool:
+        """Whether `adapter_name` is still merged into the base weights of any `_lora_loadable_modules`
+        component, used to keep the pipeline-wide `_merged_adapters` bookkeeping in sync with the real,
+        per-component PEFT merge state after a partial `unfuse_lora(components=...)` call."""
+        for component in self._lora_loadable_modules:
+            model = getattr(self, component, None)
+            if model is None or not issubclass(model.__class__, (ModelMixin, PreTrainedModel)):
```
Comment (BenjaminBossan) [inline src/diffusers/loaders/lora_base.py]: Why `issubclass(model.__class__, ...)` and not `isinstance(model, ...)`?

File: tests/lora/utils.py [hunk 4]
```diff
@@ -1711,6 +1711,45 @@ def test_simple_inference_with_text_lora_denoiser_fused_multi(
...
+        self.assertTrue(
+            pipe.num_fused_loras == 1,
+            f"adapter-1 is still merged into {denoiser_component_name}, but num_fused_loras dropped to "
+            f"{pipe.num_fused_loras=}, {pipe.fused_loras=}",
+        )
+        self.assertIn("adapter-1", pipe.fused_loras)
+
+        # Now unfuse the remaining component; bookkeeping should correctly drop to 0.
+        pipe.unfuse_lora(components=[denoiser_component_name])
+        self.assertTrue(pipe.num_fused_loras == 0, f"{pipe.num_fused_loras=}, {pipe.fused_loras=}")
```
Comment (BenjaminBossan) [inline tests/lora/utils.py:1751]: For completeness, I would also check "adapter-1" is not in `pipe.fused_loras`.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Generally looks good and makes sense. Just some small comments.

> unfuse_lora's existing docstring doesn't mention that _merged_adapters/num_fused_loras is pipeline-wide rather than per-component

Yes, that's worth documenting.

File: src/diffusers/loaders/lora_base.py (hunk 3 above)
Comment (ErenAta16) [reply src/diffusers/loaders/lora_base.py]: That's copied from the existing pattern right above it in the same unfuse_lora loop (and used throughout this file, 16 occurrences), not a deliberate choice on my part - happy to switch to isinstance() here if you'd rather not perpetuate it, but left it matching the surrounding code for now.

File: tests/lora/utils.py (hunk 4 above)
Comment (ErenAta16) [reply tests/lora/utils.py:1751]: Added in f6c54bc - now also asserts adapter-1 is out of fused_loras once it's unfused from every component.

**Comment by ErenAta16** [comment]
Pushed f6c54bc addressing the review: added the pipeline-wide vs per-component note to unfuse_lora's docstring, and replied inline on the two other points.

File: src/diffusers/loaders/lora_base.py (hunk 3 above)
Comment (BenjaminBossan) [reply src/diffusers/loaders/lora_base.py]: I see. IMO, `isinstance` makes more sense but maybe I'm missing something. I'll leave that decision up to the Diffusers maintainers.

File: src/diffusers/loaders/lora_base.py [hunk 5]
```diff
@@ -635,6 +635,11 @@ def unfuse_lora(self, components: list[str] | None = None, **kwargs):
             unfuse_text_encoder (`bool`, defaults to `True`):
                 Whether to unfuse the text encoder LoRA parameters. If the text encoder wasn't monkey-patched with the
                 LoRA parameters then it won't have any effect.
+
+        Note that `num_fused_loras`/`fused_loras` (backed by `self._merged_adapters`) are tracked pipeline-wide,
```
Comment (BenjaminBossan) [inline src/diffusers/loaders/lora_base.py]: Let's not mention implementation details here, as those don't matter to the user. It's sufficient to mention the part about `fused_loras`, as this is what's exposed to the uesr.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for the changes, just one nit.

**Review by sayakpaul — APPROVED** [review]
Thanks for your work!

File: src/diffusers/loaders/lora_base.py (hunk 5 above)
Comment (sayakpaul) [reply src/diffusers/loaders/lora_base.py]: I think it's fine to keep it as is and it can be replaced library-wide to fix on a common pattern.

**Comment by sayakpaul** [comment]
@bot /style

File: src/diffusers/loaders/lora_base.py (hunk 5 above)
Comment (ErenAta16) [reply src/diffusers/loaders/lora_base.py]: Fair, trimmed in b52aa28 - just the fused_loras behavior now, dropped the _merged_adapters mention.

File: src/diffusers/loaders/lora_base.py (hunk 3 above)
Comment (ErenAta16) [reply src/diffusers/loaders/lora_base.py]: Sounds like the consensus is to leave it as-is for now and revisit library-wide later, so no change here.
